modules/videoio/src/cap_libcamera.cpp

```diff
@@ -0,0 +1,1001 @@
+                {
+                    if (options_->verbose)
+                        CV_LOG_INFO(NULL, "VIDEOIO(Libcamera): Requests created");
+                    return;
+                }
+                std::unique_ptr<Request> request = camera_->createRequest();
+                if (!request)
+                    CV_Error(cv::Error::StsAssert, "failed to make request");
+                requests_.push_back(std::move(request));
+            }
+            else if (free_buffers[stream].empty())
+                CV_Error(cv::Error::StsAssert, "concurrent streams need matching numbers of buffers");
+
+            FrameBuffer *buffer = free_buffers[stream].front();
+            free_buffers[stream].pop();
+            if (requests_.back()->addBuffer(stream, buffer) < 0)
+                CV_Error(cv::Error::StsAssert, "failed to add buffer to request");
+        }
+    }
+}
+
+void LibcameraCapture::requestComplete(Request *request)
+{
+    if (request->status() == Request::RequestCancelled)
+        return;
+
+    CompletedRequest *r = new CompletedRequest(sequence_++, request);
+    CompletedRequestPtr payload(r, [this](CompletedRequest *cr)
+                                { this->queueRequest(cr); });
+    {
+        std::lock_guard<std::mutex> lock(completed_requests_mutex_);
+        completed_requests_.insert(r);
+    }
+
+    // We calculate the instantaneous framerate in case anyone wants it.
+    uint64_t timestamp = payload->buffers.begin()->second->metadata().timestamp;
+    if (last_timestamp_ == 0 || last_timestamp_ == timestamp)
+        payload->framerate = 0;
+    else
+        payload->framerate = 1e9 / (timestamp - last_timestamp_);
+    last_timestamp_ = timestamp;
+
+    msg_queue_.Post(Msg(MsgType::RequestComplete, std::move(payload)));
+}
+
+void LibcameraCapture::configureDenoise(const std::string &denoise_mode)
+{
+    using namespace libcamera::controls::draft;
+
+    static const std::map<std::string, NoiseReductionModeEnum> denoise_table = {
+        {"off", NoiseReductionModeOff},
+        {"cdn_off", NoiseReductionModeMinimal},
+        {"cdn_fast", NoiseReductionModeFast},
+        {"cdn_hq", NoiseReductionModeHighQuality}};
+    NoiseReductionModeEnum denoise;
+
+    auto const mode = denoise_table.find(denoise_mode);
+    if (mode == denoise_table.end())
+        CV_Error(cv::Error::StsAssert, "Invalid denoise mode " + denoise_mode);
+    denoise = mode->second;
+
+    controls_.set(NoiseReductionMode, denoise);
+}
+
+LibcameraCapture::LibcameraCapture() : LibcameraCapture(-1) {}
+
+LibcameraCapture::LibcameraCapture(int camera_index)
+{
+    options_ = std::unique_ptr<Options>(new Options());
+    controls_ = controls::controls;
+    if (!options_)
+        options_ = std::make_unique<Options>();
+    controls_.clear();
+    options = static_cast<Options *>(GetOptions());
+    still_flags = FLAG_STILL_NONE;
+    options->photo_width = 2560;
+    options->photo_height = 1440;
```

Comment:
I propose to set minimal resolution to be sure the VideoCapture opens with any camera. I would say, 640x480 should be enough as for preview.